lib/scanner/walk.go

```diff
@@ -70,16 +72,28 @@ type Config struct {
 	UseWeakHashes bool
 }
 
-type CurrentFiler interface {
-	// CurrentFile returns the file as seen at last scan.
-	CurrentFile(name string) (protocol.FileInfo, bool)
+type haveWalker interface {
+	// Walk passes all local file infos from the db which start with prefix
+	// to out and aborts early if ctx is cancelled.
+	Walk(prefix string, ctx context.Context, out chan<- *protocol.FileInfo)
 }
 
-func Walk(ctx context.Context, cfg Config) chan protocol.FileInfo {
+type fsWalkResult struct {
+	path string
+	info fs.FileInfo
+	err  error
+}
+
+type ScanResult struct {
+	New *protocol.FileInfo
```

Comment:
Having pointers to FileInfos here makes my spider sense tingle that we're doing something clever with updating them and having another pointer somewhere else that sees the change. I don't think that's the case right? If not I think we should go with just the structs here. There's an argument to be made that the struct is growing and that copying it everwhere is not super efficient, but it's not huge and we do that everywhere else...

Same comment on the chan types where you send FileInfos.

Reply:
The idea is that delete is new = nil, old = something and new discovery is vica versa.

Reply:
Actually, I did it a little bit different: `New` is never `nil`, but always a valid file info just as before (i.e. for a deleted file, the deleted flag is set, version increased, blocks wiped, ...).  Only if no file existed previously `Old` is `nil`. So I could make these structs instead of pointers and have a "default" struct in that case and `Old` is identifiable as "didn't exist before" by having an empty version.